add otel telemetry doc (#10977)
istio/istio.io: content/en/docs/tasks/observability/logs/otel-provider/index.md

---

> title: OpenTelemetry Provider

Reviewer: ```suggestion
title: OpenTelemetry
```

Author: fixed.

> description: This task shows you how to configure Envoy proxies to print access logs with OpenTelemetry Provider.

Reviewer: ```suggestion
description: This task shows you how to configure Envoy proxies to send access logs to an OpenTelemetry collector.
```

Author: fixed.

weight: 10
keywords: [telemetry,logs]
aliases:
    - /docs/tasks/telemetry/access-log

>     - /docs/tasks/telemetry/logs/access-log/

Reviewer: are these correct?  this is a new doc, right? i think we can skip `aliases` ?

Author: yes. fixed

owner: istio/wg-policies-and-telemetry-maintainers
test: yes
---

After completing this task, you understand how to have your application participate in [Envoy's access logging](https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage) with [OpenTelemetry (gRPC) Access Log](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/access_loggers/open_telemetry/v3/logs_service.proto).

> Envoy proxies send access information to [otel-collector](https://github.com/open-telemetry/opentelemetry-collector)'s standard output.

Reviewer: ```suggestion
In this example, the Envoy proxies send access logs to an [OpenTelemetry collector](https://github.com/open-telemetry/opentelemetry-collector), which is configured to print the logs to standard output.
```

Author: update

> The standard output of otel-collector's containers can then be printed by the `kubectl logs` command.

Reviewer: ```suggestion
The standard output of the OpenTelemetry collector can then be accessed via the `kubectl logs` command.
```

{{< boilerplate before-you-begin-egress >}}

{{< boilerplate start-httpbin-service >}}

{{< boilerplate start-otel-collector-service >}}

## Enable Envoy's access logging

> Istio offers a few ways to enable access logs. Use of the Telemetry API is recommended

Reviewer: WDYT about not mentioning the other ways at all?

Maybe: "To enable access logging, use the Telemetry API."  with a link to the Telemetry API docs ?

Author: sound good

### Using Telemetry API

The Telemetry API can be used to enable or disable access logs:

> Run the following script to edit the `MeshConfig`, update the YAML files to add `otel` provider.

Reviewer: ```suggestion
Edit `MeshConfig` to add an OpenTelemetry provider, named `otel`.  This will involve adding an extension provider stanza like:

    extensionProviders:
    - name: otel
      envoyOtelAls:
        service: otel-collector.istio-system.svc.cluster.local
        port: 4317

The final configuration should look something like:
```

{{< text bash >}}

> $ cat <<EOF | kubectl apply -n istio-system -f -

Author: this can be improved if we set `otel` provider by default, but I'm not sure that's the right way. @douglas-reid WDYK?

Reviewer: I'm working on a proposal that would eliminate the need to add services to MeshConfig. If that is accepted, it would mean that this step would be replaced by adding a label to the `otel-collector.istio-system` service, which should be a better user experience.

I think we should pursue that (or similar), rather than trying to figure out a one-size-fits-all default provider with a universal service address.

Author: actually, I hope we can move all service endpoints out of `MeshConfig`.
In the past, ALS require a static cluster, after https://github.com/envoyproxy/envoy/pull/19193 this's not needed anymore.

> $ cat <<EOF | kubectl apply -n istio-system -f -

Reviewer: looking at this again, I think we should avoid trying to apply a new ConfigMap. who knows what we may accidentally overwrite.

maybe just highlight the new lines and suggest `kubectl edit cm/istio` ?

Author: I'm not sure how to achieve that in a doc test.

Reviewer: Can we have the doc tests do this, but have the doc itself describe a more flexible process ?

Reviewer 2: > Can we have the doc tests do this, but have the doc itself describe a more flexible process ?

We have coded tests differently in some places. In this case the docs would talk about a more general way to make the change, and the test is just a specific implementation, which should be OK.

Author: > Can we have the doc tests do this, but have the doc itself describe a more flexible process ?

can we add a `otel` ExtesionProvider in dmeo profile, just like the `envoy` do? I think we that would be useful when adding e2e test.

Reviewer: > can we add a `otel` ExtesionProvider in dmeo profile, just like the `envoy` do? I think we that would be useful when adding e2e test.

I'm open to that, as long as we supply an `addon` yaml for generic otel collector.

Author: there's something [here](https://github.com/istio/istio/tree/master/samples/open-telemetry)

apiVersion: v1
kind: ConfigMap
metadata:
  name: istio
  namespace: istio-system
data:
  mesh: |-
    accessLogFile: /dev/stdout
    defaultConfig:
      discoveryAddress: istiod.istio-system.svc:15012
      proxyMetadata: {}
      tracing:
        zipkin:
          address: zipkin.istio-system:9411
    enablePrometheusMerge: true
    extensionProviders:
    - name: otel
      envoyOtelAls:
        service: otel-collector.istio-system.svc.cluster.local
        port: 4317
    rootNamespace: istio-system
    trustDomain: cluster.local
  meshNetworks: 'networks: {}'
EOF
{{< /text >}}

> 

Reviewer: I think we should add a sentence that is essentially a second step. "Add a Telemetry resource that tells Istio to send access logs to the OpenTelemetry collector."

spec:
  meshConfig:
    accessLogFile: /dev/stdout
    extensionProviders:
    - name: otel
      envoyOtelAls:
        service: otel-collector.istio-system.svc.cluster.local
        port: 4317
    defaultProviders:
      accessLogging:
      - envoy
      - otel
{{< /text >}}

Otherwise, add the equivalent setting to your original `istioctl install` command, for example:

{{< text syntax=bash snip_id=none >}}
$ istioctl install -f <your-istio-operator-config-file>
{{< /text >}}

## Default access log format

Istio will use the following default access log format if `accessLogFormat` is not specified:

{{< text plain >}}
[%START_TIME%] \"%REQ(:METHOD)% %REQ(X-ENVOY-ORIGINAL-PATH?:PATH)% %PROTOCOL%\" %RESPONSE_CODE% %RESPONSE_FLAGS% %RESPONSE_CODE_DETAILS% %CONNECTION_TERMINATION_DETAILS%
\"%UPSTREAM_TRANSPORT_FAILURE_REASON%\" %BYTES_RECEIVED% %BYTES_SENT% %DURATION% %RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)% \"%REQ(X-FORWARDED-FOR)%\" \"%REQ(USER-AGENT)%\" \"%REQ(X-REQUEST-ID)%\"
\"%REQ(:AUTHORITY)%\" \"%UPSTREAM_HOST%\" %UPSTREAM_CLUSTER% %UPSTREAM_LOCAL_ADDRESS% %DOWNSTREAM_LOCAL_ADDRESS% %DOWNSTREAM_REMOTE_ADDRESS% %REQUESTED_SERVER_NAME% %ROUTE_NAME%\n
{{< /text >}}

> The following table shows an example using the default access log format for a request sent from `sleep` to `httpbin`:

Reviewer 2: On my computer, this table is very wide and overwrites the right hand column.

Reviewer 2: Other than the table, this LGTM.

Reviewer 2: Other than the table, this LGTM.

Author: this table come from another page, and I do not change anything, I have no idea about what's happen on your compute.
can we merged this first, let's see what happen to https://preliminary.istio.io/?

Reviewer 3: https://deploy-preview-10977--preliminary-istio.netlify.app/latest/docs/tasks/observability/logs/otel-provider/

I'm OK with merging it and fixing the rendering bug separately.  (I don't think you've done anything wrong.)

Author: the page looks good.

Reviewer 2: Here's an example on what I see. As Craig mentioned, we can fix this rendering later.
<img width="1635" alt="Screen Shot 2022-03-15 at 7 57 12 AM" src="https://user-images.githubusercontent.com/10537847/158382651-1fd1aaf6-9d7a-4096-bf94-bd460caa1d78.png">

Author: this's a common issue, see https://istio.io/latest/docs/tasks/observability/logs/access-log/
I can take a look at this later this week.

---
title: OpenTelemetry
description: This task shows you how to configure Envoy proxies to send access logs with OpenTelemetry collector.
weight: 10
keywords: [telemetry,logs]
owner: istio/wg-policies-and-telemetry-maintainers
test: yes
---

> After completing this task, you understand how to have your application participate in [Envoy's access logging](https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage) with [OpenTelemetry (gRPC) Access Log](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/access_loggers/open_telemetry/v3/logs_service.proto).
> In this example, the Envoy proxies send access logs to an [OpenTelemetry collector](https://github.com/open-telemetry/opentelemetry-collector), which is configured to print the logs to standard output.
> The standard output of the OpenTelemetry collector can then be accessed via the `kubectl logs` command.

Reviewer 3: ```suggestion
The Envoy proxies can be configured to export their [access logs](https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage) in [OpenTelemetry format](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/access_loggers/open_telemetry/v3/logs_service.proto).
In this example, the proxies send access logs to an [OpenTelemetry collector](https://github.com/open-telemetry/opentelemetry-collector), which is configured to print the logs to standard output.
The standard output of the OpenTelemetry collector can then be accessed via the `kubectl logs` command.
```

{{< boilerplate before-you-begin-egress >}}

{{< boilerplate start-httpbin-service >}}

{{< boilerplate start-otel-collector-service >}}

## Enable Envoy's access logging

To enable access logging, use the [Telemetry API](/docs/tasks/observability/telemetry/).

> ### Using Telemetry API
> 
> The Telemetry API can be used to enable or disable access logs:

Reviewer 3: ```suggestion
```

Assuming this is the only supported method, we don't need the heading for the choice

> Edit `MeshConfig` to add an OpenTelemetry provider, named `otel`.  This will involve adding an extension provider stanza like:

Reviewer 3: ```suggestion
Edit `MeshConfig` to add an OpenTelemetry provider, named `otel`.  This involves adding an extension provider stanza:
```

{{< text yaml >}}
extensionProviders:
- name: otel
  envoyOtelAls:
    service: otel-collector.istio-system.svc.cluster.local
    port: 4317
{{< /text >}}

The final configuration should look something like:

{{< text bash >}}
$ cat <<EOF | kubectl apply -n istio-system -f -
apiVersion: v1
kind: ConfigMap
metadata:
  name: istio
  namespace: istio-system
data:
  mesh: |-
    accessLogFile: /dev/stdout
    defaultConfig:
      discoveryAddress: istiod.istio-system.svc:15012
      proxyMetadata: {}
      tracing:
        zipkin:
          address: zipkin.istio-system:9411
    enablePrometheusMerge: true
    extensionProviders:
    - name: otel
      envoyOtelAls:
        service: otel-collector.istio-system.svc.cluster.local
        port: 4317
    rootNamespace: istio-system
    trustDomain: cluster.local
  meshNetworks: 'networks: {}'
EOF
{{< /text >}}

> Add a Telemetry resource that tells Istio to send access logs to the OpenTelemetry collector.

Reviewer 3: ```suggestion
Next, add a Telemetry resource that tells Istio to send access logs to the OpenTelemetry collector.
```
